File: 4.1-calculos-matematicos/python/main.py
import time
import random
import js  # type: ignore
import gc  # type: ignore
import sys  # type: ignore
from pyscript import display  # type: ignore


def create_matrix(size):
    return [[random.random() for _ in range(size)] for _ in range(size)]


def get_memory_usage_py():
    gc.collect()
    return sum(sys.getsizeof(obj)
               for obj in gc.get_objects()) / (1024 * 1024)


def multiply_matrices(size):
    A = create_matrix(size)
    B = create_matrix(size)
    C = [[0] * size for _ in range(size)]

    start = time.time()

    for i in range(size):
        for j in range(size):
            _sum = 0
            for k in range(size):
                _sum += A[i][k] * B[k][j]
            C[i][j] = _sum
    end = time.time()

    #   Execution time

    execution_time = (end - start) * 1000
    display(
        f"PyScript (Listas nativas): {execution_time:.3f} ms", target="pyscript-output")

    #   RAM
    memory_usage = get_memory_usage_py()
    display(f"Uso de memoria: {memory_usage:.3f} MB", target="pyscript-output")

    js.endTimerWebAssembly()


def run_py_benchmark(event):
    multiply_matrices(200)

File: 4.1-calculos-matematicos/javascript/main.js
function runNodeBenchmark() {
    fetch("http://localhost:3000/")
        .then(response => response.json())
        .then(data => {
            let outputDiv = document.getElementById("nodeJs-output");

            let titleDiv = document.createElement("div");
            titleDiv.textContent = "Node.js (Servidor)";
            outputDiv.appendChild(titleDiv);

            let timeDiv = document.createElement("div");
            timeDiv.textContent = `Tiempo de Ejecucion: ${data.executionTime} ms`;
            outputDiv.appendChild(timeDiv);

            let cpuDiv = document.createElement("div");
            cpuDiv.textContent = `Uso de CPU: ${data.cpuUsage} %`;
            outputDiv.appendChild(cpuDiv);

            let memoryDiv = document.createElement("div");
            memoryDiv.textContent = `Uso de Memoria: ${data.memoryUsage} MB`;
            outputDiv.appendChild(memoryDiv);

        })
        .catch(error => console.error("Error:", error));
}

function createMatrix(size) {
    let matrix = new Array(size);
    for (let i = 0; i < size; i++) {
        matrix[i] = new Array(size);
        for (let j = 0; j < size; j++) {
            matrix[i][j] = Math.random();
        }
    }
    return matrix;
}

function multiplyMatrices(size) {
    let A = createMatrix(size);
    let B = createMatrix(size);
    let C = new Array(size).fill(0).map(() => new Array(size).fill(0));

    let start = performance.now();
    for (let i = 0; i < size; i++) {
        for (let j = 0; j < size; j++) {
            let sum = 0;
            for (let k = 0; k < size; k++) {
                sum += A[i][k] * B[k][j];
            }
            C[i][j] = sum;
        }
    }
    let end = performance.now();

    let resultTime = Number((end - start).toFixed(2));
    let result = `JavaScript (Arrays nativos): ${resultTime} ms`;
    let resultDiv = document.createElement("div");
    resultDiv.textContent = result;
    document.getElementById("javascript-output").appendChild(resultDiv);
}


function getMemoryUsageJS() {
    if (performance.memory) {
        let memoryUsed = performance.memory.usedJSHeapSize / (1024 * 1024); // Convertir a MB
        return `Memoria usada en JavaScript: ${memoryUsed.toFixed(2)} MB`;
    } else {
        return "Medición de memoria no soportada en este navegador.";
    }
}

function runJSBenchmark() {
    multiplyMatrices(200);

    let memoryDiv = document.createElement("div");
    memoryDiv.textContent = getMemoryUsageJS;
    document.getElementById("javascript-output").appendChild(memoryDiv);
}